File: CP/LeetCode/Python/6.py
# 6. Zigzag Conversion

from __future__ import annotations


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        rows = [""] * min(numRows, len(s))
        row = 0
        down = False
        for c in s:
            rows[row] += c
            if row in [0, numRows - 1]:
                down = not down
            row += 1 if down else -1
        return "".join(rows)

File: CP/LeetCode/Java/6.java
// 6. Zigzag Conversion

class Solution {
    public String convert(final String s, final int numRows) {
        if (numRows == 1) {
            return s;
        }

        int totalRows = Math.min(numRows, s.length());
        StringBuilder[] rows = new StringBuilder[totalRows];
        for (int index = 0; index < totalRows; index = index + 1) {
            rows[index] = new StringBuilder();
        }

        int currentRow = 0;
        boolean movingDown = false;

        char[] characters = s.toCharArray();
        for (int index = 0; index < characters.length; index = index + 1) {
            char currentChar = characters[index];
            rows[currentRow].append(currentChar);

            if (currentRow == 0 || currentRow == numRows - 1) {
                movingDown = !movingDown;
            }

            if (movingDown) {
                currentRow = currentRow + 1;
            } else {
                currentRow = currentRow - 1;
            }
        }

        StringBuilder finalResult = new StringBuilder();
        for (int index = 0; index < totalRows; index = index + 1) {
            finalResult.append(rows[index]);
        }

        return finalResult.toString();
    }

    public static void main(final String[] args) {
        Solution solution = new Solution();
        String output = solution.convert("PAYPALISHIRING", 3);
        System.out.println(output); // Output: PAHNAPLSIIGYIR
    }
}
